Implement TCP transport reqchannel retry logic

lukasraska opened the pull request:
### What does this PR do?
Implements retry mechanism for TCP transports ReqChannel as it's currently done for ZeroMQ.
 
### What issues does this PR fix or reference?
Fixes #59162

### Previous Behavior
Once first publish fails, event is lost forever

### New Behavior
Event publish is retried multiple times

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltstack.com/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [ ] Docs
- [x] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

lukasraska committed:
Implement TCP transport reqchannel retry logic

lukasraska commented:
FYI @waynew.
I would love to get your insight into the test implementation, the currently implemented integration test requires setting fixed `return_retry_timer` in default conf files in order to have it at least a bit deterministic and I haven't found an easy option how to set specific opts for single test (except creating new fixture with own minion factory).

Thanks

lukasraska committed:
Use custom Salt Minion factory for tests, increase the run/wait time

lukasraska commented:
re-run pr-ubuntu1604-py3-pytest

mattp- commented:
nice 👍

lukasraska commented:
Should probably be merged after #59172 , so there is no conflict with tests

s0undt3ch commented on tests/pytests/integration/test_minion.py:
```diff
@@ -0,0 +1,68 @@
+            break
+
+    assert salt_minion_retry.id in data
+    assert data[salt_minion_retry.id] is True
```
If you're just doing a `test.sleep` and a `jobs.lookup_jib` you shouldn't need that many tweaks to the minion config...
Just do:
```python
@pytest.fixture(scope="function")
def salt_minion_retry(salt_master_factory):
    # override the defaults for this test
    config_overrides = {
        "return_retry_timer_max": 0,
        "return_retry_timer": 60,
    }
    factory = salt_master_factory.get_salt_minion_daemon(
        random_string("retry-minion-),
        config_overrides=config_overrides,
        extra_cli_arguments_after_first_start_failure=["--log-level=debug"],
    )
    factory.register_after_terminate_callback(
        pytest.helpers.remove_stale_minion_key, salt_master_factory, factory.id
    )

    with factory.started():
        yield factory
```

lukasraska replied:
Thanks, updated

s0undt3ch commented on tests/pytests/integration/test_minion.py:
```diff
@@ -0,0 +1,68 @@
+    with salt_master.stopped():
+        # verify we don't yet have the result and sleep
+        assert salt_run_cli.run("jobs.lookup_jid", jid, _timeout=60).json == {}
+        time.sleep(70)
```
Why are we sleeping this much?
Because `return_retry_timer` is `60`? Can't we make `return_retry_timer` lower?

lukasraska replied:
Yeah, because retry interval is 60, so to ensure the first event return was missed. I've actually had it set to lower values (10 seconds) and it ran fine on my laptop. But once it executes in Jenkins test suite, it gets very flaky, because it takes time to start the master.

And since there is default tries = 3, we only have one/two shots at the event return (#59237 addresses this, when/if it's merged, this test could perform much better), so one minute seems to work well with Jenkins as well.

s0undt3ch replied:
Hmm. I'm not happy we have a test that sleeps that long, our test suite is already slow.... But if that's what's needed to remove flakiness. So be it.
Could we add a comment explaining why it's set that high then?

lukasraska replied:
Yeah, I'm not happy with the test slowness as well. When one of these two PRs gets merged, I will try to incorporate the test changes so it's slower. Meanwhile I've added explanation why it's set that high.

s0undt3ch replied:
Thanks

s0undt3ch commented:
Also, fyi, #59172 has been merged/

lukasraska committed:
Simplify minion factory fixture

lukasraska committed:
Add comment describing the need of long sleep in test

s0undt3ch commented on tests/pytests/integration/test_minion.py:
```diff
@@ -0,0 +1,52 @@
+import time
```
Let's instead add this test module under `tests/pytests/integration/minion/test_return_retries.py`

s0undt3ch replied:
Or publish retries, whichever makes more sense.

lukasraska replied:
Sure thing, moved it there. I've kept the `test_return_retries` name as it's generally named as `return` in docs/code.

lukasraska committed:
Rename test_minion to minion/test_return_retries

lukasraska commented:
re-run pr-amazon2-py3-pytest

mattp- commented:
bump. any reason not to merge this?

lukasraska committed:
Update test_return_retries to rely on return_retry_tries config

lukasraska commented:
I expect due to the upcoming CVE release the merging has been postponed.

Anyway, I've rebased it, so it's up-to-date with master, updated the test with larger `return_retry_tries` (as the PR for that has been merged, so it could work much better without such large sleep) and changed the slowTest decorator to pytest mark, FYI @s0undt3ch

lukasraska commented on tests/pytests/integration/minion/test_return_retries.py:
```diff
@@ -0,0 +1,52 @@
+        yield factory
+
+
+@slowTest
```
Already commited, pre-commit hook is already replacing it

s0undt3ch replied:
yeah, this was an old review that I finished today :)

lukasraska committed:
Use contextlib.suppress instead of except/pass

lukasraska commented on tests/pytests/unit/transport/test_tcp.py:
```diff
@@ -249,3 +249,118 @@ def test_tcp_pub_server_channel_publish_filtering_str_list(temp_salt_master):
+        salt_message_client.timeout_message(message_id_2, message_body_2)
+        raise future_new.exception()
+    except salt.exceptions.SaltReqTimeoutError:
+        pass
```
Pushed

lukasraska commented:
re-run pr-centos7-py3-pytest

lukasraska commented:
re-run pr-freebsd122-py3-pytest

lukasraska commented:
re-run pr-centos7-py3-pytest
